### ISSUE TITLE
proposal: `sort_interfaces`

### ISSUE THREAD

**Issue by scheglov** [dart-lang/linter#4546]
## sort_interfaces

*Replace <lint_name> above with your proposed rule name.  (See notes on naming in [Writing Lints].)*

## Description

Sort interfaces alphabetically.

## Details

**BAD:**
```dart
class I1 {}
class I2 {}
class A implements I2, I1 {}
```

**GOOD:**
```dart
class I1 {}
class I2 {}
class A implements I1, I2 {}
```

## Kind

Style advice.

## Bad Examples

See above.

## Good Examples 

See above.

## Discussion

https://dart-review.googlesource.com/c/sdk/+/312349/4/pkg/analyzer/lib/dart/element/element.dart

The order of interfaces does not have any semantics, and in the analyzer team we prefer sorting such such things, to remove subjective ordering, and make it consistent.

### Discussion checklist

- [ ] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [ ] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).

- [ ] If there's any prior art (e.g., in other linters), please add references here.
  We have a few other "sort_X" lints.

- [ ] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [ ] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by bwilkerson** [comment]
@natebosch @munificent

**Comment by natebosch** [comment]
Removing subjectivity and increasing consistency SGTM.

**Comment by lrhn** [comment]
I usually order interfaces in order of importance, 

If a class implements both `Foo` and `Comparable<Foo>`, the `Foo` seems more important.
Alphabetical ordering is arbitrary.

If some one wants the lint, that's for them, but I wouldn't want it as part of core or recommended lints. Or want it myself.

It has no correlation with code quality, at least not a positive one.

**Comment by pq** [comment]
> It has no correlation with code quality, at least not a positive one.

I think @natebosch has the upsides right:

> Removing subjectivity and increasing consistency SGTM.

We have plenty of lints that aren't so much about quality per se but rather about consistency. The pay-off comes in easing contributions (and code-reviews).

I say **+1**.

**Comment by natebosch** [comment]
> I usually order interfaces in order of importance.

I wonder if you are an outlier or not - if that is a common approach I agree that enforcing an arbitrary ordering could be a net negative.

On the whole I expect this lint to be relatively low impact - it's not common to have a long list of interfaces, and the short lists are relatively easy to read even in shuffled order IMO.

**Comment by lrhn** [comment]
I did a quick regexp based extract of multi-implements clauses from the SDK repo skd/lib/ and pkg/ directories:
https://gist.github.com/lrhn/87ea295f0ec3df1cf149db47ff1c7e33
(Some of those are from tests. I tried to filter out all 3-letters-or-less or ends-in-digit names, but there are definitely some test types left.)

I'd say that sorting by importance is the rule, if there is one.

When there are multiple related classes, it's common to put their shared interfaces last, and the ones where they differ first. 
Examples:
```
implements Int8List, TrustedGetRuntimeType {
implements Uint16List, TrustedGetRuntimeType {
```
Something like `ToJsonable` is always last, whether the previous interface is called `Message` or `Worker`.

So there is definitely a separation between *specific* and *general* interfaces in families of types, with specific ones coming before general ones.

**Comment by munificent** [comment]
> I usually order interfaces in order of importance.

Ascending or descending? :) My heuristic descending: I list interfaces that are more important to the identity or meaning of the class first followed by interfaces that are less essential or describe capabilities that it has. I have no idea how widespread that approach is.

I ran [a script](https://gist.github.com/munificent/c874c8e7efd03990ca075c78ef06051f) on a big corpus of pub packages and open source Flutter code:

```
-- Kind (214266 total) --
 171915 ( 80.234%): No implements clause       ===========================
  39841 ( 18.594%): Only one interface         =======
   1655 (  0.772%): Not in alphabetical order  =
    855 (  0.399%): In alphabetical order      =
```

In my own code, I definitely wouldn't want a lint to alphabetize them.

**Comment by scheglov** [comment]
OK, thanks everyone for the discussion.
Thank you for the statistics @munificent
I think we don't want to do this.
